packages/safe-ds-vscode/src/extension/pythonServer.ts

```diff
@@ -0,0 +1,334 @@
+            modulePath = '';
+        }
+        //
+        if (!codeMap.hasOwnProperty(modulePath)) {
+            codeMap[modulePath] = {};
+        }
+        const content = generatedDocument.getText();
+        codeMap[modulePath]![sdsNoExtFilename] = content;
+    }
+    executionInformation.set(id, {
+        generatedSource: lastGeneratedSource,
+        sourceMappings: new Map<string, BasicSourceMapConsumer>(),
+        path: pipelinePath,
+        source: lastExecutedSource,
+        calculatedPlaceholders: new Map<string, string>(),
+    });
+    sendMessageToPythonServer(
+        createProgramMessage(id, {
+            code: codeMap,
+            main: {
+                modulepath: '', // TODO maybe change generator: path.relative(workspaceRoot, path.dirname(documentUri.fsPath)).replaceAll('/', '.').split(".").concat(mainPackage).filter(str => str.length !== 0).join("."),
+                module: mainModuleName,
+                pipeline: mainPipelineName,
+            },
+        }),
+    );
+};
+
+/**
+ * Send a message to the python server using the websocket connection.
+ *
+ * @param message Message to be sent to the python server. This message should be serializable to JSON.
+ */
+export const sendMessageToPythonServer = function (message: PythonServerMessage): void {
+    const messageString = JSON.stringify(message);
+    logOutput(`Sending message to python server: ${messageString}`);
+    pythonServerConnection!.send(messageString);
+};
+
+const manageSubprocessOutputIO = function (process: child_process.ChildProcessWithoutNullStreams) {
+    process.stdout.on('data', (data: Buffer) => {
+        logOutput(`[Runner-Out] ${data.toString().trim()}`);
+    });
+    process.stderr.on('data', (data: Buffer) => {
+        logOutput(`[Runner-Err] ${data.toString().trim()}`);
+    });
+    process.on('close', (code) => {
+        logOutput(`[Runner] Exited: ${code}`);
+    });
+};
+
+const connectToWebSocket = async function (): Promise<void> {
+    // Attach WS
+    return new Promise<void>((resolve, reject) => {
+        const tryConnect = function () {
+            pythonServerConnection = new WebSocket(`ws://127.0.0.1:${pythonServerPort}/WSMain`, {
+                handshakeTimeout: 10 * 1000,
+            });
+            pythonServerConnection.onopen = (event) => {
+                pythonServerAcceptsConnections = true;
+                logOutput(`[Runner] Now accepting connections: ${event.type}`);
+                resolve();
+            };
+            pythonServerConnection.onerror = (event) => {
+                if (event.message.includes('ECONNREFUSED')) {
+                    logOutput(`[Runner] Server is not yet up. Retrying...`);
+                    setTimeout(tryConnect, 50);
+                    return;
+                }
+                logError(`[Runner] An error occurred: ${event.message} (${event.type}) {${event.error}}`);
+                reject();
+            };
```

Comment:
The startup logic is a little too optimistic ("it'll work eventually"). Some ideas for refinement:

1. Exponentially increase the timeout (e.g. 50ms, 100ms, 200ms, 400ms, 800ms). Only retry 5 times or so. If it's still not working then, the server is unlikely to come up.
1. Try again if the user updates the `safe-ds.runner.command` setting ([`onDidChangeConfiguration`](https://code.visualstudio.com/api/references/vscode-api)) using the logic from 1.
1. If the startup is triggered after the launch or a settings change, just fail silently. The extension should work fine for editing, even if the user did not install the runner.
1. If the user triggers the run command, try to start the runner again, if it's not up, using the logic from 1. If it cannot be started in this case, show an error message (IIUC, [`showErrorMessage`](https://code.visualstudio.com/api/references/vscode-api#window.showErrorMessage) is what you need here).

Reply:
The startup now includes the new runner functionality and a method to get the version is called before starting the server (could be used eventually as a compatibility check)